### ISSUE TITLE
Organise fud busting material

### ISSUE THREAD

**Issue by Czino** [Bitcoin-FUD/btcfudbuster#1]
First challenge is to organise fudbusting material.

I identified following types of material:

- url
- image
- video
- plain text

All 4 types need to be representable in the long run, for sake of getting off the ground quickly, I propose to start with urls first.

Open question I have is, how do we connect keywords (e.g. environment) with the material. Maybe we tag each item with keywords and possibly choose a random link, image, etc.. if multiple apply?

**Comment by pox** [comment]
Do we want the bot to respond when triggered, like with tweet_stamp bot? Or do we want it to actively search for terms and try to guess the topic? I think the former is probably better - you don't want the bot to chime in on false-positives (it'll make it seem stupid). I'm asking because this affects the organization of the material (tags, keywords, etc).

URLs are a good start, I agree. Most also have an image that twitter previews (but not all).

In terms of organizing, I suggest using a simpler, more human-friendly format like TOML/YAML, to encourage contributions from more people and not just devs. 

Also suggest setting up a GitHub Action to validate the material. e.g. we don't want the link/text to be longer than N characters. basically a unit test that runs when someone submits a PR. happy to help set it up if needed.

**Comment by Czino** [comment]
Yes, according to the twitter API rules, a bot should only respond to mentions and the bot needs to tag the user as well.

Text text should indeed have a max length. That's 280 characters. A url always has 23 characters (24 with a space beforehand). We also need to discount the length of two usernames (the one who mentioned the bot and the one who should read the bot's answer)

Usernames can be long let's say 16 characters per username to be on the safe side + space and @ symbol, so 36 characters in total.

So we have 280-24-36=220 characters.

I will have a deeper look into TOML/YAML. Anything that makes contribution easier, I understand that JSON is maybe not the most non-dev friendly format.

**Comment by Czino** [comment]
I think TOML looks user friendlier, there's an npm package to parse it into json so that'd be perfect for us.

**Comment by Czino** [comment]
TOML example
```
tags="energy, energy waste, energy consumption, electricity, scorch the earth, boil the oceans"
answer="Indeed #Bitcoin uses a lot of energy, but that's not a bad thing. Check out this article..."
links=[
  "https://nakamotoinstitute.org/mempool/bitcoin-does-not-waste-energy/",
  "https://hbr.org/2021/05/how-much-energy-does-bitcoin-actually-consume",
  "https://blog.trezor.io/three-myths-about-bitcoins-energy-consumption-ef613a1f3d5?gi=c24d8988688e",
  "https://docsend.com/view/adwmdeeyfvqwecj2",
  "https://www.upstreamdata.ca/post/natural-gas-venting-how-bitcoin-solved-a-160-year-old-problem",
  "https://bitcoinandenergy.medium.com/bitcoin-mining-and-the-case-for-more-energy-90094ce25fac",
  "https://www.coindesk.com/the-last-word-on-bitcoins-energy-consumption",
  "https://danhedl.medium.com/pow-is-efficient-aa3d442754d3"
]
```

**Comment by pox** [comment]
Yes, TOML is ported to every possible language and is very type-safe, so it's a good choice.
But notice the above would have to be nested inside some array. Something like this:

```TOML
[[]]
answer = "Indeed #Bitcoin uses a lot of energy, but that's not a bad thing. Check out this article..."
links = [
    "https://danhedl.medium.com/pow-is-efficient-aa3d442754d3",
]
tags = 'energy, energy waste, energy consumption, electricity, scorch the earth, boil the oceans'

[[]]
answer = "Indeed #Bitcoin uses a lot of energy, but that's not a bad thing. Check out this article..."
links = [
    'https://danhedl.medium.com/pow-is-efficient-aa3d442754d3',
]
tags = 'china, china ban'
```

Still arguably easier than JSON.

> Yes, according to the twitter API rules, a bot should only respond to mentions and the bot needs to tag the user as well.

The biggest challenge is likely to be correctly deducing the relevant FUD. Even something as straight forward as the above energy example could look really silly if someone tweets a response to: `Bitcoin doesn't use that much energy, compared to the fiat system it replaces.` and the bot replies with `Indeed #Bitcoin uses a lot of energy`.

I'm not trying to pick on this particular answer, of course. Just trying to illustrate a point that the answer really should "understand" the FUD it is responding to. The existence of the word `energy` by itself might not be enough. 

Another example: `I don't have the energy to explain to you why Bitcoin has no intrinsic value`
Response: `Indeed #Bitcoin users a lot of energy... ` 😁

Perhaps the right approach isn't trying to do NLP but rather rely on the person mentioning the bot to specify the topic.
e.g. `@btcfudbuster teach this fool about #energy please` with the hashtag servings as the unambiguous topic identifier. *and* to kinds of put a ribbon on the whole thing, if someone doesn't mention a hashtag topic the bot could reply with some generic response listing the topics it knows about. e.g.

> Beep boop, I am @btcfudbuster 🤖
> If you mention me and a topic hashtag I will reply with a relevant article.
> Topics: #energy #china #PoS #bans #privacy

**Comment by Czino** [comment]
`[[]]` throws an error when I use a TOML parser (http://binarymuse.github.io/toml-node/) is it standard syntax?
I thought we could have one file for each entry and let a compiler run when code is deployed to pack it into a single JSON or a database (depends how much the data grows)

> Another example: I don't have the energy to explain to you why Bitcoin has no intrinsic value
>Response: Indeed #Bitcoin users a lot of energy... 😁

Yes, that's going to be a tricky one.

The way the prototype is currently works: it scans each word of the tweet and matches them against the tags. Yes we have a few entries with #energy but we will also have some about. #intrinsic and #value. So that entry should be preferred.
Of course that will not eliminate false positive.
If we could teach the bot text understanding, that would be ideal but also requires proper training.

>Perhaps the right approach isn't trying to do NLP but rather rely on the person mentioning the bot to specify the topic. e.g. @btcfudbuster teach this fool about #energy please

Yes, that would then indeed be a better solution but requires the person who summons the bot to use correct tags, which might not be a huge hurdle. The generic answer for when no tags are found could indeed suffice 👍

**Comment by pox** [comment]
One file per topic is an excellent idea! Simplifies things significantly.

Trying to implement NLP is quite tricky. Most twitter bots I know rely on rather rigid syntax. At most I know some that parse date-time durations, and that's typically implemented by pre-existing libraries, since it's such a common task. Deducing FUD topics is much harder. Even for a human there might be ambiguity. A tweet that mentions china could be about mining centralizing pressures or it could be about government bans. I don't like a bot's chances with this stuff.

Maybe the right approach could be to start simple. No NLP, just look for a hashtag and reply with the above "usage" response if not found. Then at a later stage, and crucially **once the bot has collected many real world examples** a smarter deduction engine could be implemented. Even then I wouldn't try to do manual NLP. It could actually be easier to use AWS SageMaker or something to teach a neural network how to classify text, given that we have a large enough training dataset. Should be dead simple, it's just a classifier, and we would have thousands of data by then.

WDYT?

**Comment by Czino** [comment]
Alright, I implemented more material in the toml format.

I fully agree kicking off with a rigid syntax based on tags and later on when having a large enough training set expand to this scheme.

I am still undecided wether to use hashtags or scanning the tweet of the person who spreads fud or asked the question. We could try to find real world examples and backtest them.

There's also a query.mjs file to test tags or tweets and see what the result would be.

**Comment by Czino** [comment]
I am thinking, why not scanning the summoning users hashtags and give them more weight (e.g. 5 but also scan the previous tweet for keywords with less weight (e.g 0.5)?

**Comment by pox** [comment]
> I am still undecided wether to use hashtags or scanning the tweet of the person who spreads fud or asked the question.

The problem with scanning the tweet is it's so ambiguous. It's basically naive NLP. The hashtags solve the problem by making the topic explicit, so I think it's key for accuracy. But yea, once we see a few thousand of them we can train a model to deduce the hashtag.

> I am thinking, why not scanning the summoning users hashtags and give them more weight (e.g. 5 but also scan the previous tweet for keywords with less weight (e.g 0.5)?

It's an interesting idea. My personal preference is to always Keep It Simple ™️ though.

**Comment by pox** [comment]
I want to propose a (arguably) simpler way to organize the articles in `material`.

If we assume the lookup is by hashtag (at least initially) the directory structure can be:

```
material/
    energy.toml
    worthless.toml
    ...
```

Where the format is `<topic hashtag>.toml`.

**Comment by Czino** [comment]
I am in favor of keeping it easy but with the examples we already have, it will get a little difficult keeping one hashtag per file

Take a look at the folders `energy` and `mining`. I prefer having the option to give multiple hashtags to each issue. Of course the less hashtags we use the less false positives we create.

About using hashtags. Yes, let's start off by only reading hashtags from the summoning tweet, e.g. 

`@btcfudbuster teach this fool about #energy #waste please`
`@btcfudbuster does #visa use less #electricity?`
`@btcfudbuster is bitcoin only used by #criminals?`
`@btcfudbuster is bitcoin #illegal?`

**Comment by pox** [comment]
I like the approach as well. I mean, single hashtag is best if we can come up with a fitting one for each topic. Because why try to guess what the mentioning tweet is referring to if the author can tell us exactly? So if we can come up with like 7-8 hashtags that are a single word each it makes everything much simpler (and you could still catalog any given article with multiple tags, as I did in endthefud). 
But having the option of using multiple tags is rather clever. It means we don't need to come up with a huge list of hashtags, because combinations can do the work of triangulating on the most relevant ones.

Also, I really like the way you phrased the mentioning tweets as questions. It's a lot less toxic than my "teach this fool" attitude :)
